packages/langium/test/grammar/type-system/inferred-types.test.ts

```diff
@@ -578,6 +578,135 @@ describeTypes('inferred types that are used by the grammar', `
+        const grammarServices = createLangiumGrammarServices(EmptyFileSystem).grammar;
+        const document = await parseHelper<Grammar>(grammarServices)(`
+            interface Symbol {}
+            interface SuperMemberAccess extends MemberAccess {}
+
+            Expression:
+                PrimaryExpression | MemberAccess
+            ;
+
+            PrimaryExpression:
+                BooleanLiteral
+            ;
+
+            MemberAccess infers Expression:
+                PrimaryExpression (
+                    {SuperMemberAccess.receiver=current} '.' member=[Symbol:'super'] |
+                    {infer MemberAccess.receiver=current} '.' member=[Symbol:'foo']
+                )+
+            ;
+
+            BooleanLiteral:
+                {infer BooleanLiteral} value?='true' | 'false'
+            ;
+        `);
+
+        expect(document.parseResult.lexerErrors).toHaveLength(0);
+        expect(document.parseResult.parserErrors).toHaveLength(0);
+
+        checkTypes(document.parseResult.value);
+    });
+
+    function checkTypes(grammar: Grammar) {
+        const sortByName = (a: {name: string}, b: {name: string}) => (a.name?.localeCompare(b.name));
+        const toSubstring = (o: {toString: () => string}) => {
+            // this specialized 'toString' function uses the default 'toString' that is  producing the
+            //  code generation output, and strips everything not belonging to the actual interface/type declaration
+            const sRep = o.toString();
+            return sRep.substring(
+                0, 1 + (sRep.includes('interface') ? sRep.indexOf('}') : Math.min(sRep.indexOf(';') ))
+            );
+        };
+
+        const { parserRules, datatypeRules } = collectAllAstResources([grammar]);
+        const { interfaces: inferredInterfaces, unions } = collectInferredTypes(Array.from(parserRules), Array.from(datatypeRules));
+
+        const unionsString = unions.map(toSubstring).join('\n').trim();
+        expect(unionsString).toBe(s`
+            export type Expression = MemberAccess | PrimaryExpression;
+            export type PrimaryExpression = BooleanLiteral;
+        `);
+
+        const inferredInterfacesString = inferredInterfaces.sort(sortByName).map(toSubstring).join('\n').trim();
+        expect(inferredInterfacesString).toBe(s`
+            export interface BooleanLiteral extends AstNode {
+                readonly $container: MemberAccess;
+                value: boolean
+            }
+            export interface MemberAccess extends AstNode {
+                readonly $container: MemberAccess;
+                member: Reference<Symbol>
+                receiver: PrimaryExpression
+            }
+        `);
+
+        const allInterfaces = collectAst(undefined!, [grammar]).interfaces;
+        const allInterfacesString = allInterfaces.sort(sortByName).map(toSubstring).join('\n').trim();
+        expect(allInterfacesString).toBe(s`
```

Comment:
Suggestion: I'm not sure whether this kind of assertion will survive the next refactoring of the type inference. Can you refactor the string based comparison to the one that's used in the rest of the tests? What do you think about that?

Reply:
Basically I like the string-based assertion checking a lot, and I explicitly advocate that.

Relying on the "generated code"-based `toString()` representation can of course be questioned.
It was just convenient and easy to understand, once one is familiar the generated AST representation, I currently have the time to implement a dedicated text rep.

I suggest to implement a beneficial `toString()` rep of the internal type objects during the refactoring,
and suggest to consider the string based comparison an inspiration, as I personally prefer the human readable diffs in case of failures a lot over a certain diff in
```
function expectProperty(interfaceType: InterfaceType, property: Property | string): void {
    if (typeof property === 'string') {
        const prop = getProperty(interfaceType, property)!;
        expect(prop).toBeDefined();
        expect(prop.optional).toStrictEqual(false);
    } else {
        const prop = getProperty(interfaceType, property.name)!;
        expect(prop).toBeDefined();
        expect(prop.optional).toStrictEqual(property.optional);
        expect(prop.typeAlternatives.length).toStrictEqual(property.typeAlternatives.length);
        for (let i = 0; i < prop.typeAlternatives.length; i++) {
            const actualType = prop.typeAlternatives[i];
            const expectedType = property.typeAlternatives[i];
            expect(actualType.types).toEqual(expectedType.types);
            expect(actualType.array).toEqual(expectedType.array);
            expect(actualType.reference).toEqual(expectedType.reference);
        }
    }
}
```
where first have to investigate the context of the failure.